code/ZAS/Zone.dm

```diff
@@ -199,6 +208,40 @@ Class Procs:
 	//for(var/turf/T in unsimulated_contents)
 //		to_chat(M, "[T] at ([T.x],[T.y])")
 
+// There are some behaviors that we want to happen upon certain thresholds. For instance, if slippery ice is spread across many turfs that melts upon a certain temperature,
+// we want to know when the zone's temperature is raised above that threshold.
+// In an ideal non-BYOND environment we could use something like events to hookup each ice tile to. However, our in-house implementation of events suck for performance. So you
+// can just hardcode the check here.
+/zone/proc/check_for_events()
+	// Only initialize ice_puddle_list and do these checks after cryotheum has been introduced to the zone.
+	if( ice_puddle_list != null )
+		for(var/obj/effect/overlay/puddle/ice/existing_ice in ice_puddle_list)
+			existing_ice.current_temp = air.temperature
+		// Below freezing and cryotheum in the air, create ice.
+		if( air.temperature <= T0C && air.molar_density(GAS_CRYOTHEUM) > MOLES_CRYOTHEUM_VISIBLE / CELL_VOLUME )
+			for(var/turf/simulated/T in contents)
+				if(isemptylist(T.search_contents_for(/obj/effect/overlay/puddle/ice)))
```

Comment:
This essentially does a search through contents on EVERY turf in order to see if there's a puddle on it, and if not, then spawn a new one.
Pretty inefficient, right? More efficient would be to add var/puddle to every turf so that we can do if(turf.puddle == null) in very fast speed, but I was told adding vars to turf is a bad idea and I agree.
Any other way?

Reply:
Faster to use `locate()` or a typed loop. Don't know which is faster nowadays.
```
if(!(locate(/obj/effect/whatever) in T))
```
Or
```
var/found = FALSE
for(var/obj/effect/whatever/varname in T)
	found = TRUE
	break
if(!found)
```
The first is certainly more elegant. I don't know which one performs better.

Reply:
The second one made zone/tick total CPU become 4% of what /datum/controller/matser/proc/RunQueue's was during a cryotheum flood. First one was 7%. Before it was 14%.
Not sure what % of that 4% was other zone/tick stuff but seems to work well.